### PR TITLE
Added internal ExceptionMustBeRethrown helper to Target

### PR DESCRIPTION
**Pull request by 304NotModified**
- easier to use
- better context capturing
- Better error config scoping (LogFactory instead of LogManager)

Discussion: should we make this internal? (as private protected is not yet possible)

if liked, could also make one for LayoutRenderer and Layout

### CODE REVIEW

File: src/NLog/Internal/ExceptionHelper.cs [hunk 1]
```diff
@@ -78,8 +78,9 @@ public static bool IsLoggedToInternalLogger(this Exception exception)
         /// </summary>
         /// <param name="exception">The exception to check.</param>
         /// <param name="loggerContext">Target context of the exception.</param>
+        /// <param name="logFactory">For checking <see cref="LogFactory.ThrowExceptions"/> and <see cref="LogFactory.ThrowConfigExceptions"/>. Falls back to <see cref="LogManager"/> </param>
         /// <returns><c>true</c>if the <paramref name="exception"/> must be rethrown, <c>false</c> otherwise.</returns>
-        public static bool MustBeRethrown(this Exception exception, IInternalLoggerContext loggerContext = null)
+        public static bool MustBeRethrown(this Exception exception, IInternalLoggerContext loggerContext = null, LogFactory logFactory = null)
```
Comment (snakefoot) [inline src/NLog/Internal/ExceptionHelper.cs:83]: Maybe add LoggingConfiguration as property to `IInternalLoggerContext`, and then extract the LogFactory from that?

File: src/NLog/Internal/ExceptionHelper.cs [hunk 2]
```diff
@@ -99,8 +100,10 @@ public static bool MustBeRethrown(this Exception exception, IInternalLoggerConte
                     InternalLogger.Log(exception, level, "Error has been raised.");
             }
 
+            logFactory = logFactory ?? LogManager.LogFactory;
```
Comment (snakefoot) [inline src/NLog/Internal/ExceptionHelper.cs:103]: Think to make it non-breaking then you should both check local `logFactory` AND global `LogManager.LogFactory`

File: src/NLog/Internal/ExceptionHelper.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Internal/ExceptionHelper.cs:83]: OK will do!

File: src/NLog/Internal/ExceptionHelper.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/Internal/ExceptionHelper.cs:103]: OK, So I think

- logfactory.ThrowConfigExceptions, if null, then fallback to LogManager.ThrowConfigExceptions
- logfactory.ThrowExceptions, if false, then fallback to LogManager.ThrowExceptions? (unsure about this one)

Comment (snakefoot) [reply src/NLog/Internal/ExceptionHelper.cs:103]: If one of them is `true` then it should throw and explode.

Comment (304NotModified) [reply src/NLog/Internal/ExceptionHelper.cs]: I'm not sure if we should keep it the old behavior, is kinda bug that the properties on LogFactory aren't used.

Comment (snakefoot) [reply src/NLog/Internal/ExceptionHelper.cs]: Yes "bug" should be fixed with NLog 5.0 where lot of static behavior is being fixed.

**Comment by 304NotModified** [comment]
maybe move this to NLog 5 then for the breaking changes?

**Comment by snakefoot** [comment]
Where is the breaking change?


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


maybe move this to NLog 5 then for the breaking changes?

**Comment by 304NotModified** [comment]
this https://github.com/NLog/NLog/pull/3953#discussion_r423172916

**Comment by snakefoot** [comment]
> this #3953 (comment)

As it is like this, then I don't see it as breaking:

```c#
if (logFactory?.ThrowExcetions==true || LogManager.ThrowExceptions)
{
}
```

**Comment by 304NotModified** [comment]
it. hurts. 

I've added this because in tests I still need to change the logmanager while I never use logmanager. But OK, that's a NLog 5 thing. Will add a fat todo

File: src/NLog/Internal/ExceptionHelper.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/Internal/ExceptionHelper.cs]: Updated it!

**Comment by 304NotModified** [comment]
@snakefoot please double check (the last commit should be enough)

**Comment by snakefoot** [comment]
Looks good to me.

**Comment by 304NotModified** [comment]
Unfortunately Travis is still unstable sometimes:

image

**Comment by snakefoot** [comment]
> Unfortunately Travis is still unstable sometimes:

Guess more of those Wait-forever calls needs to be removed, or maybe Travis-machines are just unstable.

**Comment by 304NotModified** [comment]
> Guess more of those Wait-forever calls needs to be removed, or maybe Travis-machines are just unstable.

Could be the latter, as .NET isn't officially supported. Maybe move to devops for Linux/Mono also

**Comment by snakefoot** [comment]
Find that Azure Devops is worse than Travis and AppVeyor :)

**Comment by 304NotModified** [comment]
Why? It has tons of more features and really a better GUI

**Comment by snakefoot** [comment]
Stability wise.

**Comment by 304NotModified** [comment]
True, we have some timeouts there also (60m).

**Pull request merged**